src/ucp/api/ucpx.h

```diff
@@ -19,7 +19,63 @@
 
 BEGIN_C_DECLS
 
+enum ucp_am_rendezvous_params_field {
+  UCP_AM_RENDEZVOUS_FIELD_IOVEC_SIZE = UCS_BIT(0)  /* Max size of the iovec returned by the initial AM */
+};
 
+typedef struct ucp_am_rendezvous_params {
+  uint64_t field_mask ;
+  size_t iovec_size ;
+} ucp_am_rendezvous_params_t ;
+
+
+typedef void (*ucp_am_data_function_t)(void *target, void *source, size_t bytes, void *cookie) ;
+typedef ucs_status_t (*ucp_am_local_function_t)(void *arg, void *cookie, ucp_dt_iov_t *iovec, size_t iovec_length) ;
+
+typedef struct ucp_am_rendezvous_recv {
+    ucp_am_local_function_t local_fn ;
+    void * cookie ;
+    ucp_am_data_function_t data_fn ;
+    void *data_cookie ;
+    size_t iovec_max_length ;
+    size_t iovec_length ;
+    ucp_dt_iov_t iovec[1] ;
+  } ucp_am_rendezvous_recv_t ;
+
+typedef ucs_status_t (*ucp_am_rendezvous_callback_t)(void *arg, void *data, size_t length,
+                                          ucp_ep_h reply_ep, unsigned flags, size_t remaining_length, ucp_am_rendezvous_recv_t *recv);
+
+ucs_status_t ucp_worker_set_am_rendezvous_handler(ucp_worker_h worker, uint16_t id,
```

Comment:
What would be invoked if both callbacks (regular and rendezvous) were set?

Reply:
In my current implementation, the server is expected to set both callbacks. The ucp library can decide that it does not want to perform rendezvous processing for a particular active message call; if it does that, it converts it to a regular (eager) active message and drives the regular callback.

Reply:
maybe define a single callback like the one you proposed and declare the old one as deprecated?

Reply:
I think the existing API is fine for many purposes, also my code will drive the existing API if a particular active message call is unsuitable for driving the new API. So I don't want to deprecate the existing API. Maybe that can be the subject of a future change.